### ISSUE TITLE
Integration with iproute2

### ISSUE THREAD

**Issue by whitequark** [nushell/nushell#346]
I've wanted a structured shell for a *very* long time, and I'm very excited to see nushell as it does many of the things I like. One of the major pain points with Unix tools I've had over years is scripting network configuration: `ifconfig`/`ip`: `ip` fixed many of the worst issues with *input*, but to get any data out of it, you still have to parse it with `awk` or something like that.

It would be great to see `iproute2` integration in nushell, and I think it would also make an excellent showcase of its capabilities when it comes to system administration, much like one can show how excellent PowerShell is at that on Windows.

**Comment by rrichardson** [comment]
oh heck yes.  I have so many scripts lying around that pipe from `ip route` or `ip addr`  
It's even helpful without a plugin : 

` ip route | split-column " " | where Column1 == "192.168.122.0/24" | get "Column3"` 
`virbr0` 

:)

**Comment by jonathandturner** [comment]
This needs a fix I haven't landed just yet, but since someone mentioned that iproute2 has json support, I wanted to drop this here to add to the design conversation:

```
> ip -j route | from-json
---+----------------+---------------+--------+----------+--------+-----------+-------+--------------
 # | dst            | gateway       | dev    | protocol | metric | flags     | scope | prefsrc 
---+----------------+---------------+--------+----------+--------+-----------+-------+--------------
 0 | default        | 192.168.1.254 | wlp8s0 | dhcp     | 600    | [0 items] |       |  
 1 | 169.254.0.0/16 |               | wlp8s0 |          | 1000   | [0 items] | link  |  
 2 | 192.168.1.0/24 |               | wlp8s0 | kernel   | 600    | [0 items] | link  | 192.168.1.68 
---+----------------+---------------+--------+----------+--------+-----------+-------+--------------
```

It's not full iproute2 integration, but seems like it could be a start.

**Comment by rrichardson** [comment]
nice!   @jonathandturner  - how does one get the tables to auto-expand like that? 
I get
```
> ip -j route | from-json
-----------
 value
-----------
 [4 items]
-----------
```

**Comment by rrichardson** [comment]
OK.  I notice that where clauses don't seem to work on columns that have `null` values.  Is this the problem to which you were referring?  E.g.  
```
> ip -j route | from-json
---+------------------+-------------+---------+----------+--------+-----------+-------+---------------
 # | dst              | gateway     | dev     | protocol | metric | flags     | scope | prefsrc
---+------------------+-------------+---------+----------+--------+-----------+-------+---------------
 0 | default          | 192.168.1.1 | wlo1    | dhcp     | 600    | [0 items] |       |
 1 | 172.17.0.0/16    |             | docker0 | kernel   |        | [1 item]  | link  | 172.17.0.1
 2 | 192.168.1.0/24   |             | wlo1    | kernel   | 600    | [0 items] | link  | 192.168.1.189
 3 | 192.168.122.0/24 |             | virbr0  | kernel   |        | [1 item]  | link  | 192.168.122.1
---+------------------+-------------+---------+----------+--------+-----------+-------+---------------

```

```
> ip -j route | from-json | where gateway != ""
error: Could not evaluate (MissingProperty)
- shell:1:32
1 | ip -j route | from-json | where gateway != ""
  |                                 ^^^^^^^^^^^^^^^ could not evaluate
```

I was able to run a `where` filter on any column where every row has a value, but I cannot run on any row without a all values filled.  

Is there a ticket in for this?  If not, I'lll create it.   I did a quick search and didn't see any issues relating to this.

**Comment by jonathandturner** [comment]
@rrichardson - yeah that might be a bug. Yes, please file a ticket for that, it should work.

**Comment by rrichardson** [comment]
IMO, when #419 is fixed, this could be closed.  With the help of the `-j` flag, I think we can have everything we need from iproute2.

**Comment by thegedge** [comment]
@rrichardson I think I'll close this one now, to tidy up. I think we should fix #419 for the next release.

In the meantime, one could make an alias:

```
alias routes [] { ip -j route | from json | default gateway "" | default metric "" }
```

(and any other columns to default)
